New Python SE3 interface

rbridges12 opened the pull request:
## Summary
Addresses Issue: #19 

This is a refactor of the SE3 class and addition of an SO3 class, moving it to the interface we discussed. Before this is merged in, there are a few questions I want to get some input on:
- should we add equality overrides? so we can check `se3_1 == se3_2` or `so3_1 == so3_2`?
- should we enforce a normalization invariant for quaternions, or just leave it to the user? if invariant, where should this be enforced? (just constructor, setter function, etc)
- should we stick with the current "dataclass" style of accessing data where you just do `se3.position` to get position, or is there a benefit in adding getter functions so we can do `se3.get_position()`?
- should we check numpy array sizes in constructors and functions?
- should we copy numpy arrays in constructors instead of just assigning them directly?
- where do we want to handle error checking for TF lookups in the drive code? There are several places in the drive code that now are vulnerable to an SE3 being returned as None, are these a concern?

What features did you add, bugs did you fix, etc? 
- Changed SE3 interface
- changed SE3 underlying implementation from ROS Transform message type to numpy array and SO3 object
- added SO3 object to represent rotations
- refactored all drive code that previously used the old SE3 interface to use this new one

### Did you add documentation to the wiki?
No, there is already a wiki entry on Poses and the SE3/SO3 groups, which mentions this class. I could add a further entry on how to use this class if there is interest.

## How was this code tested? 
I tested this on unit tests that I wrote (more info below) and in the sim. In sim, I used navigation.launch to run a full search spiral which successfully found the AR tag, and to run a sequence of 3 non AR tag waypoints, which was also successful. Since the changes to actual nav code were very minor, I think this is good enough.

### Did you test this in sim? 
Yes

### Did you test this on the rover?
No

### Did you add unit tests? 
Yes, I wrote unit tests providing decent coverage of every function in the SE3 and SO3 classes. I used numbers from [this quaternion calculator](https://www.andre-gaschler.com/rotationconverter/) as well as Wolfram Alpha. All the non TF involved functions were written purely using the python unittest framework, and the TF functions were tested using a rostest file in order to run some transform publishers. As noted in a TODO comment in the SE3 test file, I haven't yet tested `publish_to_tf_tree()` with a non static transform, because this will probably require a more complex setup.

niwhsa9 reviewed (changes requested):
Took a first pass at it. Looks really good, and I like the thorough unit testing. Here are my thoughts on your questions:

> should we add equality overrides? 

sure, I'd implement it by adding an `is_approx(se3 : SE3, tolerance)` function in each class, then the equality override calls the function with a default tolerance.  

> should we enforce a normalization invariant for quaternions, or just leave it to the user? 

I think its good to have these be true dataclasses, and so lets just leave it to the user. Just make a big ol' note of it in the docstring and the wiki that this class doesn't enforce normalization but expects it to function properly. 

> should we stick with the current "dataclass" style of accessing data...

Yes . This is most idiomatic in Python.

> should we check numpy array sizes in constructors and functions?

No, its bloated. Just assume caller competence. 

> should we copy numpy arrays in constructors instead of just assigning them directly?

Doing this as a `@dataclass` will solve this problem

> where do we want to handle error checking for TF lookups in the drive code? There are several places in the drive code that now are vulnerable to an SE3 being returned as None, are these a concern?

don't perform exception handling in library code. Its better to handle exceptions as close to the caller level as possible imo. Raise them to the caller.

rbridges12 commented on src/util/SE3.py:
```diff
@@ -1,102 +1,93 @@
+
+    def publish_to_tf_tree(
+        self,
+        tf_broadcaster: tf2_ros.TransformBroadcaster | tf2_ros.StaticTransformBroadcaster,
```
mypy seems happy with it, altho this is allegedly a python 3.10+ feature. I used it because I thought it was more readable

rbridges12 commented on src/util/SE3.py:
```diff
@@ -1,102 +1,93 @@
+        try:
+            tf_msg = tf_buffer.lookup_transform(parent_frame, child_frame, rospy.Time()).transform
+
+        except (
```
idk I'm kinda torn on this, on one hand I agree that it might be stylistically better to explicitly catch errors higher up rather than checking for None, but on the other hand this saves us from writing a large try except statement every time you look up something in the TF tree (which we do a lot). Its also worth noting that this is a very common exception for us to run into since it takes a short amount of time for TFs to be published and accessible. Also, I feel like returning None actually makes sense and isn't just a convoluted error checker, because nothing has gone catastrophically wrong, you just looked for a TF and it wasn't there so you get nothing.

niwhsa9 replied:
if nothing else, by returning None you've completely obfuscated the underlying error. Not every caller will want to handle a `LookupException`, `ConnectivityException` and `ExtrapolationException` the same way. e.g. Perhaps in some applications a lookup exception is worth a retry, but a connectivity exception needs to be a soft quit.

rbridges12 replied:
Ok fair enough, ig we just have to deal with a ton of try except blocks. So you think I should just get rid of the try catch block in `from_tf_tree()` and add it back to the places I removed it from the drive code?

rbridges12 replied:
also should I add some :raises: tags to the docstring for each of these exceptions explaining what they are?

niwhsa9 replied:
yep sg on both

rbridges12 replied:
Done, now the nav code should be effectively back where it was before this PR. There are a lot of places where these errors aren't caught at all and we need to add try except statements in the nav code, but it's a non trivial addition that will require thinking about it carefully so I'll open an issue for someone to do later

rbridges12 commented:
> > should we copy numpy arrays in constructors instead of just assigning them directly?
> 
> Doing this as a `@dataclass` will solve this problem
Do you mean that using the `__init__` function generated by the dataclass decorator will solve this problem? If so, this doesn't seem to be true from my testing. Just to be clear, this is the problem I'm talking about:
```py
>>> x = np.array([1, 2, 3])
>>> p = SE3(position=x)
>>> p.position
array([1, 2, 3])
>>> p.position[0] = -1
>>> x
array([-1,  2,  3])
```
and from this example it doesn't seem to work:
```py
>>> @dataclass
... class Test:
...     arr: np.ndarray
...
>>> a = np.array([1, 2, 3])
>>> t = Test(a)
>>> t.arr
array([1, 2, 3])
>>> t.arr[0] = 7
>>> a
array([7, 2, 3])
```

niwhsa9 commented:
Sorry, I completely misread your original question. Yes, you can do np.copy() so there's no accident ownership transfer

rbridges12 commented on src/util/SO3.py:
```diff
@@ -0,0 +1,72 @@
+        q1 = self.quaternion
+        q2 = r.quaternion
+        q_diff = quaternion_multiply(quaternion_inverse(q1), q2)
+        angle = 2 * np.arccos(q_diff[3])
```
Added some unit tests that check this, turns out the original didn't actually pick the shortest arc so I added this solution which seems to fix it, altho it isn't the prettiest solution

niwhsa9 replied:
seems fine, I think this is pretty much unavoidable

ankithu commented:
Personally I feel like having an quality override for anything but strict equality (or a very small epsilon) is pretty unintuitive and is going to lead to confusion on what exactly the default tolerance when people go back to read code later. I think that having a function is_within_tolerance(other pose, tolerance) is a lot more clear (or you could make it not a member function and just take two poses)

niwhsa9 commented:
Yeah that's a fair point. I think I agree.

rbridges12 commented:
Yeah that's a good point, I removed the equality operator.

niwhsa9 commented on src/util/SE3.py:
```diff
@@ -1,102 +1,100 @@
-
-        :param tf: the Transform message object
-        :returns: the created SE3 pose object
+    def __init__(self, position: np.ndarray = None, rotation: np.ndarray = None):
```
dataclass actually automatically creates a constructor for you, you can provide default args: https://docs.python.org/3/library/dataclasses.html

niwhsa9 commented:
`SE3` and `SO3` can be frozen too so they are immutable once created

rbridges12 commented:
I agree that's a good idea, I moved to using the autogenerated `__init__` method provided by the dataclass and made both classes immutable. Also I added a pyproject.toml file that acts as a config for black so we don't have to add the `--line-length=120` anymore.

niwhsa9 commented on pyproject.toml:
```diff
@@ -0,0 +1,2 @@
+[tool.black]
+line-length = 120
```
nice!

qhdwight replied:
Very nice

qhdwight replied:
Maybe we can have the github action use this file?

niwhsa9 reviewed (approved):
LGTM, nice work

qhdwight commented on src/util/SE3.py:
```diff
@@ -1,102 +1,105 @@
 
-    def x_vector(self) -> np.ndarray:
+    @classmethod
+    def from_pos_quat(cls, position: np.ndarray, quaternion: np.ndarray):
```
docstring and parameter mismatch for `quaternion`

qhdwight commented on src/util/SE3.py:
```diff
@@ -1,102 +1,105 @@
+        return np.linalg.norm(p.position - self.position)
 
-    def rot_distance_to(self, p: SE3) -> float:
+    def is_approx(self, p: SE3, tolerance=1e-8) -> bool:
```
missing docstring for `tolerance`

qhdwight commented on src/util/SO3.py:
```diff
@@ -0,0 +1,81 @@
+
+          For example:
+          >>> arr = np.array([1, 2, 3, 4])
+          >>> r = SO3(quaternion=arr.copy())
```
I like this a lot

rbridges12 replied:
😊
